clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy

```diff
@@ -0,0 +1,201 @@
+    CacheData cacheData = cacheResultBuilder.onDemand.toKeep[key]
+    cacheData ? cacheData.attributes.cacheTime >= cacheResultBuilder.startTime : false
+  }
+
+  /**
+   * Generically will move cached data from onDemand to resource (Server Group, Load Balancer, Security Group) namespace.
+   * @param objectMapper
+   * @param cacheResultBuilder
+   * @param serverGroupKey
+   */
+  void moveOnDemandDataToNamespace(ObjectMapper objectMapper, TypeReference<Map<String, List<MutableCacheData>>> typeReference, CacheResultBuilder cacheResultBuilder, String key) {
+    String cacheResults = cacheResultBuilder.onDemand.toKeep[key].attributes.cacheResults as String
+    Map<String, List<MutableCacheData>> onDemandData = objectMapper.readValue(cacheResults, typeReference)
+
+    onDemandData.each { String namespace, List<MutableCacheData> cacheDatas ->
+      cacheDatas.each { MutableCacheData cacheData ->
+        cacheResultBuilder.namespace(namespace).keep(cacheData.id).with {
+          attributes = cacheData.attributes
+          relationships = cacheData.relationships
+        }
+        cacheResultBuilder.onDemand.toKeep.remove(cacheData.id)
+      }
+    }
+  }
+
+  /**
+   * Finds all on demand cache data for a given account and region.
+   * @param providerCache
+   * @param accountName
+   * @param region
+   * @return
+   */
+  Collection<Map> getAllOnDemandCacheByRegionAndAccount(ProviderCache providerCache, String accountName, String region) {
+    Collection<String> keys = providerCache.getIdentifiers(ON_DEMAND.ns).findAll { String key ->
+      Map<String, String> parsedKey = Keys.parse(key)
+      parsedKey && parsedKey.account == accountName && parsedKey.region == region
+    }
+
+    providerCache.getAll(ON_DEMAND.ns, keys).collect { CacheData cacheData ->
+      [
+        details       : Keys.parse(cacheData.id),
+        cacheTime     : cacheData.attributes.cacheTime,
+        processedCount: cacheData.attributes.processedCount,
+        processedTime : cacheData.attributes.processedTime
+      ]
+    }
+  }
+
+  /**
+   * Builds on demand cache result based upon the cache result.
+   * @param onDemandAgentType
+   * @param cacheResult
+   * @param namespace
+   * @param key
+   * @param providerCache
+   * @return
+   */
+  OnDemandResult buildOnDemandCache(Object cachedItem
+                                    , String onDemandAgentType
+                                    , CacheResult cacheResult
+                                    , String namespace
+                                    , String key) {
+    OnDemandResult result = new OnDemandResult(sourceAgentType: onDemandAgentType, cacheResult: cacheResult, evictions: [:].withDefault { _ -> [] })
+
+    if (!cachedItem) {
+      // Evict this cached item if it no longer exists.
+      result.evictions[namespace] << key
+    }
+
+    result
+  }
+
+  /**
+   * Resolves a key that contains a wildcard but still returns a unique result.
```

Comment:
I will update it to throw an exception ... Would like to know if it isn't uniquely resolving.